tests/pushbits/main.c

```diff
@@ -76,6 +79,90 @@ echo()
 	exit(1);
 }
 
+static int
+wait_ms(size_t msec)
+{
+	struct timespec ts;
+
+	ts.tv_sec = msec / 1000;
+	ts.tv_nsec = (msec % 1000) * 1000000;
+	nanosleep(&ts, NULL);
+
+	return (0);
+}
+
+static int
+chain()
+{
+	pthread_t thread;
+	int in[2];
+	int out[2];
+	const char * msg = "ping";
+	size_t msglen = strlen(msg);
+	char * buf;
+	ssize_t r;
+
+	/* Initialize output message buffer. */
+	if ((buf = malloc(msglen)) == NULL) {
+		warnp("malloc");
+		goto err0;
+	}
+
+	/* Create socket pairs for the input and output. */
+	if (socketpair(AF_UNIX, SOCK_STREAM, 0, in)) {
```

Comment:
Draft PR so that I can ask a question; I'm missing something fundamental about sockets.

- I first attempted to use `socket()` to create two sockets, then connect them with `pushbits()`.  That didn't work because it complained about not being connected.  IOW, I tried to do:

    ```
    write(s0, ...) -> pushbits(s0, s1) -> read(s1, ...)
    ```

- I now have it working with `socketpair()`, but this feels like I have too many sockets / file descriptors floating around.  I now have:

    ```
    write(in0, ...) -> socketpair(in0, in1) -> pushbits(in1, out0) -> socketpair(out0, out1) -> read(out1, ...)
    ```

Reply:
This is correct.  For "socket" think "half of a connection"; a socketpair is a complete connection.  So with two socketpairs you can be writing into one connection, pushbits takes data out of that connection and pushes it into the other, and then you read data out of the second connection.